### DISCUSSION TITLE
Consider scraping off the ability of the sender to withdraw on behalf of the recipient

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#666] [Ideas]
## Problem

One of the [findings](https://github.com/cantinasec/review-sablier/issues/22) reported by Cantina during their 1st review of the code base has come back to bite us.

During my call with [NFTFI.com](https://nftfi.com/) yesterday to discuss a Sablier V2 integration, it became apparent that allowing the sender to withdraw makes the integration difficult.

NFTFI has an escrow smart contract where people lock up their NFTs after making a borrow against them. Unfortunately, this contract cannot react to incoming ERC-20 deposits, and it can neither detect increases in ERC-20 balances over time. Consequently, the funds would be forever lost if a sender performs a withdrawal while the NFT sits in this escrow.

It is likely that NFTFI is not the only NFT platform that cannot react to/ account for sender withdrawals.

## Solution

Scrape off the ability of the sender to withdraw on behalf of the recipient.

Of course, doing so would mean that funds are forever lost if the sender has accidentally started a stream to an address that cannot interact with Sablier, such as an exchange address. There are two ways to go about this problem:

1. We just accept it and make the warning more prominent in the create stream form (although this would also pose problems for integrations)
2. We implement a permissioned relayer module whereby the Sablier multisig can allow particular accounts to withdraw on behalf of any Sablier user

Side note: it might be worth it to perform some data crunching to see how many sender-initiated withdrawals there have been in V1 and V2.

**Comment by maxdesalle** [comment]
> We just accept it and make the warning more prominent in the create stream form (although this would also pose problems for integrations)

I think this is an acceptable tradeoff. Worth noting that for traditional transactions, the sender cannot withdraw either. Once the money is gone, the money is gone. People in crypto have grown accustomed to that, I think, so it's not a problem.

> We implement a permissioned relayer module whereby the Sablier multisig can allow particular accounts to withdraw on behalf of any Sablier user

This is a bad idea, in my opinion. Just the possibility of being able to do so via a multisig would get a lot of our users to lose trust in our protocol. Full immutability matters, and a system is only as decentralized as its most centralized component.

**Reply by andreivladbrg** [reply]
I agree with both of Max's points.

>it might be worth it to perform some data crunching to see how many sender-initiated withdrawals there have been in V1 and V2.

This would be indeed useful information.

**Reply by PaulRBerg** [reply]
> Worth noting that for traditional transactions, the sender cannot withdraw either

good point

> Once the money is gone, the money is gone. 

Things are rosier with cancelable streams: the sender can recover something (a huge part, actually) if and when they realize that the recipient address is incorrect.

> Just the possibility of being able to do so via a multisig would get a lot of our users to lose trust in our protocol

Strong disagree because the worst that could happen is to trigger a taxable event for some users, which is basically nothing compared to upgrade risk. Upgradeable contracts are commonplace. For instance, a single-digit number of people can upgrade [entire rollups](https://l2beat.com/scaling/risk).

Very few people will care about our little feature for claiming assets on behalf of Sablier users. We would explain the rationale in the docs.

It is worth noting that [some of our competitors](https://docs.streamflow.finance/help/tutorials/withdrawing-payments#automatic-withdraw) offer automatic withdrawals.

> a system is only as decentralized as its most centralized component

The protocol would still remain 100% decentralized from the point of view of asset ownership. Sablier could never take ownership of user funds.

**Reply by PaulRBerg** [reply]
Anyway, we already have a separate discussion about the 2nd point: [Withdraw on behalf of](https://github.com/sablier-labs/v2-core/discussions/39). We should continue discussing there.

**Reply by razgraf** [reply]
Agree with Max. 

I would also reconsider the "permissioned relayer". Like sure, for a special customer (e.g. an enterprise with 100s of employees) it's a great idea - as we can offer a specialized B2B service that includes automated withdrawals etc. But for the general purpose audience, that's seems like an (1) unnecessary privacy/trust concern and (2) management cost for whoever the relayer is (multisig, DAO etc.)

**Reply by PaulRBerg** [reply]
My point about upgradeable contracts/ rollups above has not been criticized/ refuted.

> I would also reconsider the "permissioned relayer"

Could you explain in more details how would this work?

> for a special customer (e.g. an enterprise with 100s of employees) it's a great idea - as we can offer a specialized B2B service that includes automated withdrawals etc.

You can have an input at the time of creating the stream that lets you choose if you want automatic withdrawals or not.

> unnecessary privacy/trust concern

Which is like a blip on the radar compared to what other kinds of trust issues are out there in web3.

[Starknet has just upgraded their network and frozen the balances of all users who did not upgrade in time](https://twitter.com/0xfoobar/status/1696909831712690590).

> management cost for whoever the relayer is (multisig, DAO etc.)

We would subsidize this, and the cost would be recovered by charging users fees in exchange for providing the service.

**Reply by PaulRBerg** [reply]
I think you are worrying too much about the trust aspect, @maxdesalle, @razgraf.

It's a really, really minor issue in the wider context of web3, and it has to be weighed against the benefits of automatic withdrawals: (i) improve the UX for users who are too lazy to learn how to operate Sablier, (ii) rescue funds for stuck users, and (ii) become a source of revenues.

**Reply by maxdesalle** [reply]
Happy to agree to disagree on that point @PaulRBerg, not convinced by the benefits and the actual need on my end.

**Reply by PaulRBerg** [reply]
We need to make a decision about this so it's not (unfortunately) a case of being able to agree to disagree.

I'm tentatively in favor of implementing a system whereby withdraw relayers can be added by governance.

**Reply by PaulRBerg** [reply]
And the more I think about it, the more important the automation aspect seems to be.

We're putting quite a bit of strain on recipients to have to use our UI to claim manually - there may be use cases where automatic withdrawals are a better fit.

**Reply by PaulRBerg** [reply]
But anyway this particular discussion is about disabling sender withdrawals.

I will follow up in https://github.com/sablier-labs/v2-core/discussions/155

**Reply by maxdesalle** [reply]
After re-reading the whole conversation here, I changed my mind! I'm in soft agreement with the relayer idea.

**Reply by razgraf** [reply]
> Could you explain in more details how would this work?

You mean the example with regular vs enterprise users? Hmm, while I don't like this solution - probably separate contracts: one with no default relayers (for the public) and one with default relayers and more control to "managers". Just an idea.

**Reply by PaulRBerg** [reply]
> I changed my mind! I'm in soft agreement with the relayer idea.

Awesome

> probably separate contracts: one with no default relayers (for the public) and one with default relayers and more control to "managers"

Yeah this would be fairly complicated.

**Comment by razgraf** [comment]
Following our conversation in the thread above:

While I would definitely not downplay the trust implications *any* controlled system would bring, I can understand that for some the majority of users this can be a good feature and potentially a backup in case of human errors (e.g. sending to an exchange wallet).

Can we, by any chance, include something like a `managed` flag, where the stream recipient has the ability to opt-in/opt-out of this default relayer support (or even all operator support it that's easier)? I'm thinking this way, if we have someone that really wants full control over their stream, they can just opt-out of any pre-configured 3rd party management.

**Reply by maxdesalle** [reply]
> Can we, by any chance, include something like a managed flag, where the stream recipient has the ability to opt-in/opt-out of this default relayer support (or even all operator support it that's easier)? I'm thinking this way, if we have someone that really wants full control over their stream, they can just opt-out of any pre-configured 3rd party management.

This is the right approach! Great idea.

**Reply by PaulRBerg** [reply]
> While I would definitely not downplay the trust implications any controlled system would bring

I would downplay them in the short and the medium-term.

Rollup risk > smart contract protocol risk

Our little feature pales in comparison with the fact that the Arbitrum and Optimism multisig can rug-pull all users.

> I can understand that for some the majority of users, this can be a good feature and potentially a backup in case of human errors

Awesome

> Can we, by any chance, include something like a managed flag

Good idea. I would just let users specify a global `disabledRelayers` mapping, which is by default set to `false`, but can be manually set to `true`.

**Comment by PaulRBerg** [comment]
Feedback provided by Jorge from [Unlockd](https://www.unlockd.finance/):

> With our V2 we don't need to disable withdrawals. Our new features solve this without any changes required but it's a bit long to explain here. Will send a voicenote in a bit

Refer to https://github.com/sablier-labs/v2-core/discussions/684#discussioncomment-7056765 for more context on what Jorge said in the voice note.

**Comment by PaulRBerg** [comment]
Closing as per the rationale given [here](https://github.com/sablier-labs/v2-core/pull/682#issuecomment-1732325144).
